plugins/mac-syphon/syphon.m

```diff
@@ -147,10 +160,12 @@ static inline void check_description(syphon_t s, NSDictionary *desc)
 	    surfaces_string.UTF8String);
 }
 
-static inline void handle_new_frame(syphon_t s,
-				    SYPHON_CLIENT_UNIQUE_CLASS_NAME *client)
+static inline void handle_new_frame(syphon_t s, SyphonOpenGLClient *client)
 {
-	IOSurfaceRef ref = [client IOSurface];
+	SyphonOpenGLImage *image = [client newFrameImage];
+	IOSurfaceRef ref = (IOSurfaceRef)CFBridgingRetain(
+		/* Rules are made to be broken... like buildings! Or people! */
+		[image performSelector:@selector(surface)]);
```

Comment:
It's either that..

Or you properly create a new category for the original class (say.. `SyphonOpenGLImage+Surface.m`) - there are canonical ways to do this:

https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html#//apple_ref/doc/uid/TP40011210-CH6-SW2

Reply:
I will admit that I don't understand how that would avoid the `performSelector` call, as `surface` is still hidden in `SyphonOpenGLImage`.
I created a category for the class, but as far as I can tell I'd still have to call `surface` via the ObjC runtime. The code does indeed look nicer but I don't think it's what you intended.
I might very well be missing something though, please apologize my ignorance here. I'll push what I did as a temporary commit here so you can see what I mean.

Looking into this more I believe that the "proper" way to do this is to create our own subclasses of `SyphonServerBase` and `SyphonImageBase`. I can probably do that if we don't want to call private APIs, but would like to know that I didn't miss something major with what you mentioned above because it does look at least like a bit of work.

Reply:
By implementing a new category for `SyphonOpenGLImage` (and importing that header) you can add a public getter method that returns the property.

As Apple's docs specify:

> If you need to add a method to an existing class, perhaps to add functionality to make it easier to do something in your own application, the easiest way is to use a category.

So precisely what we need here. You effectively "overwrite" the existing class definition with your extended one. It's a common pattern in ObjC to add functionality to existing classes (e.g. adding encoders/decoders to the `NSString` class and similar use cases).

Reply:
The result would be that you would just do `[image getSurface]` and the category would just extend via:

```
@public
- (IOSurfaceRef)getSurface
{
    return _surface;
}
```

Reply:
`_surface` would still be invisible though?

Either way it doesn't matter, I changed it to subclass `SyphonServerBase` which I believe is the real proper way anyways. This has the advantage of cutting out the OpenGL middleman, removing the need for an OpenGL context and allowing us to access the IOSurfaces in a direct way.